### PR TITLE
build(core-clp): Remove `boost::iostream` dependency.

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description

PR #445 has implemented our own memory-mapped file. With our implementation, we can now remove `boost::iostream` dependency from clp and unit tests.
This PR makes necessary changes to replace `boost` memory-mapped files with `clp::ReadOnlyMemoryMappedFile`, and remove the `boost::iostream` dependency from the cmake files of `clp`, `clg`, `clo`, `make-dictionaries-readable`, and `unitTest`. Since `boost::iostream` is still used in `clp-s` and `glt`, it is not yet fully removed from our library dependency.

# Validation performed

Ensure all unit tests passed.

## Summary by CodeRabbit

- **New Features**
	- Introduced a custom `ReadOnlyMemoryMappedFile` class for improved memory mapping functionality.
  
- **Bug Fixes**
	- Simplified error handling and resource management for memory-mapped files in the `Segment` class.

- **Tests**
	- Updated tests to utilize the new `ReadOnlyMemoryMappedFile` class, enhancing the encapsulation of memory mapping functionality. 

- **Documentation**
	- Added necessary includes for the new memory mapping class in the relevant test files.

### CODE REVIEW

**Comment by kirkrodrigues** [comment]
Do you want to try merging main to see if Centos 9 resolves the workflow failure?

File: components/core/src/clp/streaming_archive/reader/Segment.cpp [hunk 1]
```diff
@@ -73,7 +54,8 @@ ErrorCode Segment::try_open(string const& segment_dir_path, segment_id_t segment
         return ErrorCode_Failure;
     }
 
-    m_decompressor.open(m_memory_mapped_segment_file.data(), segment_file_size);
+    auto const view{m_memory_mapped_segment_file.value().get_view()};
+    m_decompressor.open(view.data(), view.size());
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/streaming_archive/reader/Segment.cpp:58]: open doesn't have return value

**Comment by LinZhihao-723** [comment]
@kirkrodrigues Yeah Centos 9 resolves the workflow failure. Can u check the commit message?

File: components/core/src/clp/streaming_archive/reader/Segment.cpp [hunk 2]
```diff
@@ -33,38 +33,19 @@ ErrorCode Segment::try_open(string const& segment_dir_path, segment_id_t segment
...
-    memory_map_params.flags = boost::iostreams::mapped_file::readonly;
-    memory_map_params.length = segment_file_size;
-    // Try to map it to the same memory location as the previous memory mapped file
-    memory_map_params.hint = m_memory_mapped_segment_file.data();
-    m_memory_mapped_segment_file.open(memory_map_params);
-    if (!m_memory_mapped_segment_file.is_open()) {
+
+    // Create read-only memory mapped file
+    m_memory_mapped_segment_file.emplace(segment_path);
+    if (false == m_memory_mapped_segment_file.has_value()) {
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/reader/Segment.cpp:48]: Shouldn't this logic be that we catch `ReadOnlyMemoryMappedFile::OperationFailed` and then print the log and return the error?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/src/clp/streaming_archive/reader/Segment.cpp (hunk 2 above)
Comment (LinZhihao-723) [reply components/core/src/clp/streaming_archive/reader/Segment.cpp]: right this is a major mistake... thanks for catching this issue. Solved in the latest commit

File: components/core/src/clp/streaming_archive/reader/Segment.cpp [hunk 3]
```diff
@@ -33,47 +35,31 @@ ErrorCode Segment::try_open(string const& segment_dir_path, segment_id_t segment
...
+        m_memory_mapped_segment_file.emplace(segment_path);
+    } catch (TraceableException const& ex) {
         SPDLOG_ERROR(
                 "streaming_archive::reader:Segment: Unable to memory map the compressed "
-                "segment with path: {}",
-                segment_path.c_str()
+                "segment with path: {}. Error: {}",
+                segment_path.c_str(),
+                ex.what()
         );
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/reader/Segment.cpp:57]: Even though we're going to switch to proper error codes soon, let's still try to do a good job of logging the error:

```cpp
        auto error_code = ex.get_error_code();
        SPDLOG_ERROR(
                "streaming_archive::reader:Segment: Unable to memory map the compressed "
                "segment with path: {}. Error: {}",
                segment_path.c_str(),
                ErrorCode_errno == error_code
                        ? std::format("{}", strerror(errno))
                        : std::format("error_code={}, {}", ex.what(), error_code)
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

```
build(core-clp): Remove `boost::iostream` dependency.
```

**Pull request merged**